lib/src/rewrite.rs

```diff
@@ -842,6 +842,185 @@ pub fn move_commits(
     })
 }
 
+pub enum DuplicateCommitsDestination {
+    /// Duplicate each commit on top of their parents or other duplicated
+    /// commits.
+    Parents,
+    /// Duplicate the roots of the target set on top of the parent commits, and
+    /// duplicate the other commits on the target set onto duplicated
+    /// commits in the target set. If `children_commit_ids` is not empty,
+    /// the `children_commit_ids` will be rebased onto the heads of the
+    /// duplicated target commits.
+    Destination {
+        parent_commit_ids: Vec<CommitId>,
+        children_commit_ids: Vec<CommitId>,
+    },
+}
+
+#[derive(Default)]
+pub struct DuplicateCommitsResult {
+    /// Map of original commit ID to newly duplicated commit.
+    pub duplicated_commits: IndexMap<CommitId, Commit>,
+    /// The number of descendant commits which were rebased onto the duplicated
+    /// commits.
+    pub num_rebased: u32,
+}
+
+/// Duplicates the given `target_commits` on their original parents, or
+/// onto a new location in the graph.
+///
+/// Commits in `target_commits` are duplicated into their new location given by
+/// `destination`. This assumes that any new children commits can be rewritten,
+/// and there will be no cycles in the resulting graph. Commits in
+/// `target_commits` should be in reverse topological order.
+pub fn duplicate_commits(
+    settings: &UserSettings,
+    mut_repo: &mut MutableRepo,
+    target_commits: Vec<CommitId>,
+    destination: DuplicateCommitsDestination,
+) -> BackendResult<DuplicateCommitsResult> {
+    if target_commits.is_empty() {
+        return Ok(Default::default());
+    }
+
+    let mut duplicated_old_to_new: IndexMap<CommitId, Commit> = IndexMap::new();
+    let mut num_rebased = 0;
+
+    match destination {
```

Comment:
Since the two match arms don't share much, do you think it's better to extract a function for each? That would at least reduce indentation a bit.

Reply:
Maybe they should be separate functions? `Destination::Parents` is conceptually quite different. `Destination::Destination` is closer to `move_commits()`. If we had `rebase --keep`, the end result would be basically the same (except for the change ids.)

Reply:
I separated this into two functions.